### PR TITLE
Parallelize precompiles generation

### PR DESCRIPTION
**Pull request by petvana**
This PR parallelizes `generate_precompile.jl`. It saves half a minute building Julia, and, thanks to using `Channel`, the order of the precompiles should be still fully deterministic.

**PR**
```
...
    JULIA usr/lib/julia/sys-o.a
Generating REPL precompile statements... 40/40
Executing precompile statements... 1750/1754
Precompilation complete. Summary:
Total ─────── 112.036051 seconds
...
```

**Master**
```
...
    JULIA usr/lib/julia/sys-o.a
Generating REPL precompile statements... 40/40
Executing precompile statements... 1748/1752
Precompilation complete. Summary:
Generation ── 121.136666 seconds 80.6719%
Execution ───  29.023100 seconds 19.3281%
Total ─────── 150.159765 seconds
...
```

The next optimization would be to read `precompile_file` on the fly, that would save several seconds extra.

### CODE REVIEW

**Comment by IanButterworth** [comment]
Seems like a good thing to do, but it could make the precompilation process harder to debug. It might be good to be able to fallback to the current serial approach for debugging purposes. I guess putting optional waits on the `@async` tasks could be a simple way to do that?

Also I think the progress printing needs figuring out. It's currently a bit chaotic and hard to interpret.

File: contrib/generate_precompile.jl [hunk 1]
```diff
@@ -245,15 +245,16 @@ function generate_precompile_statements()
...
 
     # From hardcoded statements
     for statement in split(hardcoded_precompile_statements::String, '\n')
-        push!(statements, statement)
+        push!(statements_step1, statement)
     end
 
     # Collect statements from running the script
-    mktempdir() do prec_path
+    @async mktempdir() do prec_path
```
Comment (vchuravy) [inline contrib/generate_precompile.jl:257]: These should have `errormonitor` on them so that their failure is loud

Comment (petvana) [reply contrib/generate_precompile.jl]: Thank you, should be loud now.

**Comment by petvana** [comment]
I've tried to improve the printing (link to asciinema below) and add `PARALLEL_PRECOMPILATION = false` option for serial debugging 

<a href="https://asciinema.org/a/548119?autoplay=1"></a>

**Comment by petvana** [comment]
Not sure, where the current CI problem comes from.

**Comment by IanButterworth** [comment]
Try rebasing?

File: contrib/generate_precompile.jl (hunk 1 above)
Comment (petvana) [reply contrib/generate_precompile.jl]: @vchuravy  `errormonitor` causes CI to fail for some reason. Is the current approach with `fetch(step1) == :ok || throw("Step 1 of collecting precompiles failed.")` sufficient?

**Comment by IanButterworth** [comment]
The REPL precomp generation script includes intentionally throwing errors. Perhaps that's why errormonitor etc was surfacing errors that were previously ignored?

**Comment by petvana** [comment]
> The REPL precomp generation script includes intentionally throwing errors. Perhaps that's why errormonitor etc was surfacing errors that were previously ignored?

Perhaps, but I'm unable to confirm locally.

**Comment by petvana** [comment]
Besides `errormonitor`, the PR seems ready. Suggestions for improving the progress printing welcomed.

**Comment by IanButterworth** [comment]
Looks great. 
For formatting maybe this, where `Waiting` etc. are colored and the counts for the categories are colored appropriately. Also perhaps "Basic" instead of "normal"?
```
Precompile statements (Waiting, Running, Finished): 
└ Collect 1748 (Basic: 729, REPL 40/40: 1653) => Execute 1744/1748
Precompilation complete. Summary:
Total ─────── 114.175765 seconds
Outputting sysimage file...
Output ────── 162.738189 seconds
````

File: contrib/generate_precompile.jl [hunk 2]
```diff
@@ -275,8 +295,20 @@ function generate_precompile_statements()
         push!(statements_step1, statement)
     end
 
-    println("Precompile statements (Waiting, Running, Finished)")
+    println("Collect and execute precompile statements")
```
Comment (KristofferC) [inline contrib/generate_precompile.jl:298]: Maybe this should "collecting" and "executing"?

**Comment by petvana** [comment]
New fancy printing. Running for `sys.so` to make it short.

<a href="https://asciinema.org/a/548532" target="_blank"></a>

**Comment by KristofferC** [comment]
I like it!

The raw logs in e.g. CI gets a bit spammed 

```
_bk;t=1672318930123Collecting and executing precompile statements
_bk;t=1672318930123
└ Collect (Basic: ? , REPL 0/0: ? ) => Execute ? 
└ Collect (Basic: ? , REPL 0/0: ? ) => Execute ? 
└ Collect (Basic: ◐ , REPL 0/0: ? ) => Execute ? 
└ Collect (Basic: ◐ , REPL 0/0: ◐ ) => Execute ? 
└ Collect (Basic: ◐ , REPL 0/0: ◐ ) => Execute ◐ 1 (1 failed)
└ Collect (Basic: ◓ , REPL 0/0: ◓ ) => Execute ◓ 1 (1 failed)
└ Collect (Basic: ◓ , REPL 0/0: ◓ ) => Execute ◓ 2 (1 failed)
└ Collect (Basic: ◓ , REPL 0/0: ◓ ) => Execute ◓ 3 (3 failed)
└ Collect (Basic: ◓ , REPL 0/0: ◓ ) => Execute ◓ 4 (3 failed)
└ Collect (Basic: ◑ , REPL 0/0: ◑ ) => Execute ◑ 4 (3 failed)
└ Collect (Basic: ◑ , REPL 0/0: ◑ ) => Execute ◑ 5 (3 failed)
└ Collect (Basic: ◑ , REPL 0/0: ◑ ) => Execute ◑ 6 (3 failed)
└ Collect (Basic: ◑ , REPL 0/0: ◑ ) => Execute ◑ 7 (4 failed)
└ Collect (Basic: ◑ , REPL 0/0: ◑ ) => Execute ◑ 8 (4 failed)
└ Collect (Basic: ◑ , REPL 0/0: ◑ ) => Execute ◑ 9 (4 failed)
└ Collect (Basic: ◑ , REPL 0/0: ◑ ) => Execute ◑ 10 (4 failed)
└ Collect (Basic: ◑ , REPL 0/0: ◑ ) => Execute ◑ 11 (4 failed)
└ Collect (Basic: ◑ , REPL 0/0: ◑ ) => Execute ◑ 12 (4 failed)
└ Collect (Basic: ◒ , REPL 0/0: ◒ ) => Execute ◒ 12 (4 failed)
└ Collect (Basic: ◒ , REPL 0/0: ◒ ) => Execute ◒ 13 (4 failed)
└ Collect (Basic: ◒ , REPL 0/0: ◒ ) => Execute ◒ 14 (4 failed)
└ Collect (Basic: ◒ , REPL 0/0: ◒ ) => Execute ◒ 15 (4 failed)
```

Maybe check if stdout/stderr is a TTY for fancyprint?

**Comment by KristofferC** [comment]
If we want to be consistent with `Pkg.precompile`, the spinner should just be white I think.

**Comment by IanButterworth** [comment]
Looking great. Disabling the cursor is best with a try-finally because it can be annoying to lose it if the build process fails

File: contrib/generate_precompile.jl [hunk 3]
```diff
@@ -403,28 +481,34 @@ function generate_precompile_statements()
...
+    fetch(step2) == :ok || throw("Step 2 of collecting precompiles failed.")
 
+    tot_time = time_ns() - start_time
     println("Precompilation complete. Summary:")
-    print("Generation ── "); Base.time_print(gen_time);     print(" "); show(IOContext(stdout, :compact=>true), gen_time / tot_time * 100);     println("%")
-    print("Execution ─── "); Base.time_print(include_time); print(" "); show(IOContext(stdout, :compact=>true), include_time / tot_time * 100); println("%")
     print("Total ─────── "); Base.time_print(tot_time);     println()
-
+catch e
+    rethrow()
```
Comment (IanButterworth) [inline contrib/generate_precompile.jl:509]: Nit but this is redundant, you can just try-finally
```suggestion
```

**Review by IanButterworth — APPROVED** [review]
Great!

**Comment by IanButterworth** [comment]
I realized the 4 we're reporting as failures are just comments being skipped, so (I hope you don't mind) I pushed a fix for that

**Comment by IanButterworth** [comment]
> The next optimization would be to read precompile_file on the fly, that would save several seconds extra.

It seems like `--trace-compile=$precompile_file` is populated at the end of the process. Making that more parallel might take a lot of internals work

**Comment by KristofferC** [comment]
We could flush the strean for every statement printed maybe.

**Comment by IanButterworth** [comment]
I actually think I might be wrong, and we're already flushing. Investigating..

**Comment by petvana** [comment]
> It seems like `--trace-compile=$precompile_file` is populated at the end of the process. Making that more parallel might take a lot of internals work

Actually, they are populated during the process. I've already tried a very naive approach and it seems to work.
https://github.com/petvana/julia/compare/pv/parallel-repl-statements-v2...petvana:julia:pv/parallel-repl-statements-on-the-fly-v2

**Comment by IanButterworth** [comment]
Nice. I've been trying an approach that doesn't re-read and unique the statements each loop, though it's not working yet

**Comment by petvana** [comment]
I propose to leave the asynchronous reading to a different PR because it adds one more level of complexity (and space for possible dead-locks).

**Comment by IanButterworth** [comment]
Sounds good to me

**Comment by IanButterworth** [comment]
This seems to work without re-reading the total file each time https://github.com/IanButterworth/julia/commit/5ee671ad3c40533a3b3965f8bcbce20d18014cbf?w=1

**Comment by petvana** [comment]
> This seems to work without re-reading the total file each time [IanButterworth@5ee671a?w=1](https://github.com/IanButterworth/julia/commit/5ee671ad3c40533a3b3965f8bcbce20d18014cbf?w=1)

Nice, but I'm slightly worried about `readline` if it may split the last line before it's fully written. I'd rather use `readavailable` and split manually.

**Comment by IanButterworth** [comment]
Builds are passing, but win 32 tests are reporting failure without an obvious reason. Updating from master just to check

**Comment by vtjnash** [comment]
readline will never do that (if reading from a proper stream/pipe), but readavailable will often do that

**Comment by KristofferC** [comment]
Wow, I didn't mean to merge this. I tried to merge it locally using `gh merge` and it merged it here instead 😆

**Comment by IanButterworth** [comment]
Wasn't even a squash merge. That seems to easy to do accidentally..!

**Comment by KristofferC** [comment]
I think it asked me and since I thought it was local I thought whatever. The last step said "submit" which in hind sight probably should have made me more alert...

**Comment by petvana** [comment]
> readline will never do that (if reading from a proper stream/pipe), but readavailable will often do that

It seems possible if someone flushes (that is a correct think to do). The following MWE
``` julia
t = tempname()
w = open(t, "w")
r = open(t, "r")
tr = @async begin
    for i = 1:30
        sleep(0.1)
        line = readline(r)
        line == "" && continue
        println(line)
    end
end

print(w, "a+")
flush(w)
sleep(0.3)
print(w, "b\n")
close(w)

wait(tr)
```
prints only 
```
a+
```
but it is not a full line.

**Comment by vtjnash** [comment]
A file is not a stream/pipe (unless it was created with mkfifo)

**Pull request merged**